jxl/src/features/patches.rs

```diff
@@ -333,11 +471,125 @@ impl PatchesDictionary {
                 ysize: ref_pos_ysize,
             })
         }
-        Ok(PatchesDictionary {
+
+        let mut patches_dict = PatchesDictionary {
             positions,
             blendings,
             ref_positions,
             blendings_stride,
-        })
+            num_patches: vec![],
+            sorted_patches_y0: vec![],
+            sorted_patches_y1: vec![],
+            patch_tree: vec![],
+        };
+        patches_dict.compute_patch_tree()?;
+        Ok(patches_dict)
+    }
+
+    pub fn get_patches_for_row(&self, y: usize) -> Vec<usize> {
+        let mut result = vec![];
+        if self.num_patches.len() <= y || self.num_patches[y] == 0 {
+            return result;
+        }
+
+        result.reserve(self.num_patches[y]); //  Reserve space as in C++
+
+        let mut tree_idx: isize = 0;
+        loop {
+            if tree_idx == -1 {
+                break;
+            }
+
+            // Safe access using get() and unwrap_or().  No need for the assert.
+            let node = self.patch_tree.get(tree_idx as usize).unwrap_or_else(|| {
+                // TODO(firsching): Handle panic differently?
+                panic!("Invalid tree_idx: {}", tree_idx);
+            });
+
+            if y <= node.y_center {
+                for i in 0..node.num {
+                    let p = self.sorted_patches_y0[node.start + i];
+                    if y < p.0 {
+                        break;
+                    }
+                    result.push(p.1);
+                }
+                tree_idx = if y < node.y_center {
+                    node.left_child
+                } else {
+                    -1
+                };
+            } else {
+                for i in 0..node.num {
+                    let p = self.sorted_patches_y1[node.start + i];
+                    if y >= p.0 {
+                        break;
+                    }
+                    result.push(p.1);
+                }
+                tree_idx = node.right_child;
+            }
+        }
+
+        // Ensure that the relative order of patches is preserved.
+        result.sort();
+        result
+    }
+
+    pub fn add_one_row(
+        &self,
+        row: &mut [&mut [f32]],
+        position: (usize, usize),
+        xsize: usize,
+        extra_channel_info: &[ExtraChannelInfo],
+        reference_frames: &[ReferenceFrame],
+    ) {
+        let bg = row
+            .iter_mut()
+            .map(|s| s.to_vec())
+            .collect::<Vec<Vec<f32>>>();
```

Comment:
This will allocate some buffers, not sure this would be okay... Maybe we need error checks here?

Reply:
You can use `new_with_capacity` and `.extend_from_slice()` so that the closure returns `Result<Vec<f32>, _>`, and the iterator can be collected into `Result<Vec<T>, E>`.

Reply:
I just realized this function is used by the render pipeline, which doesn't handle errors.

Should I assume the allocations are small enough to not cause errors? If not, how should one handle render errors during render pipeline calls?

ping @veluca93 too

Reply:
Yes, you should assume so. Ideally, stages in the render pipeline allocate necessary temporary buffers in advance.

Reply:
I guess this could be done, but it would be messier, and I'll leave it as a TODO.